### PR TITLE
Cr 1228 nisra redirecting

### PR DESCRIPTION
**Pull request by Paul-Joel**
# Motivation and Context
Introduces region specific redirecting/messaging

# What has changed
Adds region specific trapping - the entered UAC or chosen address must now be appropriate for the site version, ie: An NI UAC cannot be used on the /en or /cy sites, only on the /ni sites. If the user is on the wrong 'site' for the data supplied they now get a page saying that they are on the wrong site, with a link to the correct one (same principle as for Scotland), rather than the site shifting them over.
This is deemed a better solution, as the nav and footer links are for the wrong site, and a user could get incorrect help, or call the wrong contact centre

All unit tests and test data have been updated to have version for all 3 regions

Required cucumber updates were stand-alone and have already been merged.

# How to test?
Enter a N UAC on the /en/start/ version
Enter a E or W UAC on teh /ni/start/ version

# Links
https://collaborate2.ons.gov.uk/jira/browse/CR-1228

# Screenshots (if appropriate):

### CODE REVIEW

**Review by cookel2 — CHANGES_REQUESTED** [review]
Hi Paul, it looks good. I'm only requesting a change to the test data. However, I think it's also worth saying that I accidentally tested with an unlinked UAC at one point. I entered it onto the start page for England. It then prompted me for a postcode so I gave it the NI one (BT38 9PT) just to see what would happen. It then gave me the "Sorry, something went wrong" message, which makes sense but might be something that we could improve if there's time.

**Review by MartinHumphrey — CHANGES_REQUESTED** [review]
Minor comments on redundant attributes passed in context to templates.
The new files created in this PR need a newline added to the end. GitHub complaining "No new line at end of file" e.g.  common-address-in-england.html.

File: app/start_handlers.py [hunk 1]
```diff
@@ -184,41 +184,85 @@ async def post(self, request):
...
+
+@start_routes.view('/ni/start/code-for-wales/')
+class StartCodeForWales(StartCommon):
+    @aiohttp_jinja2.template('start-code-for-wales.html')
+    async def get(self, request):
+        self.setup_request(request)
+        display_region = 'ni'
+        self.log_entry(request, display_region + '/start/code-for-wales')
+
+        locale = 'en'
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/start_handlers.py:257]: locale is still needed - although the page has no text marked for translation, passing locale through also sets the html lang attribute (and possibly other settings)

File: app/start_handlers.py [hunk 2]
```diff
@@ -184,41 +184,85 @@ async def post(self, request):
...
+        page_title = 'This access code is not part of the census for Northern Ireland'
+
+        await forget(request)
+
+        return {
+            'display_region': display_region,
+            'locale': locale,
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/start_handlers.py:264]: As above

File: app/start_handlers.py [hunk 3]
```diff
@@ -184,41 +184,85 @@ async def post(self, request):
...
+
+@start_routes.view('/ni/start/code-for-england/')
+class StartCodeForEngland(StartCommon):
+    @aiohttp_jinja2.template('start-code-for-england.html')
+    async def get(self, request):
+        self.setup_request(request)
+        display_region = 'ni'
+        self.log_entry(request, display_region + '/start/code-for-england')
+
+        locale = 'en'
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/start_handlers.py:237]: As above

File: app/start_handlers.py [hunk 4]
```diff
@@ -184,41 +184,85 @@ async def post(self, request):
...
+        page_title = 'This access code is not part of the census for Northern Ireland'
+
+        await forget(request)
+
+        return {
+            'display_region': display_region,
+            'locale': locale,
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/start_handlers.py:244]: As above

File: app/common_handlers.py [hunk 5]
```diff
@@ -91,6 +91,89 @@ async def get(self, request):
...
+    and selected addresses region is W
+    """
+    @aiohttp_jinja2.template('common-address-in-wales.html')
+    async def get(self, request):
+        self.setup_request(request)
+        display_region = 'ni'
+        user_journey = request.match_info['user_journey']
+
+        page_title = 'This address is not part of the census for Northern Ireland'
+        locale = 'en'
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/common_handlers.py:164]: As above

File: app/common_handlers.py [hunk 6]
```diff
@@ -91,6 +91,89 @@ async def get(self, request):
...
+    and selected addresses region is E
+    """
+    @aiohttp_jinja2.template('common-address-in-england.html')
+    async def get(self, request):
+        self.setup_request(request)
+        display_region = 'ni'
+        user_journey = request.match_info['user_journey']
+
+        page_title = 'This address is not part of the census for Northern Ireland'
+        locale = 'en'
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/common_handlers.py:138]: As above

File: tests/test_data/address_index/uprn_wales.json [hunk 7]
```diff
@@ -0,0 +1,27 @@
+{
+    "apiVersion": "1.0.0-SNAPSHOT",
+    "dataVersion": "72",
+    "response": {
+        "address": {
+            "uprn": "10023122451",
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply tests/test_data/address_index/uprn_wales.json]: Not really a big deal, as only the Region value is used for testing, however I have updated the test data for Wales and Scotland to be representative

File: app/start_handlers.py [hunk 8]
```diff
@@ -184,41 +184,85 @@ async def post(self, request):
...
+        display_region = 'ni'
+        self.log_entry(request, display_region + '/start/code-for-wales')
+
+        locale = 'en'
+        page_title = 'This access code is not part of the census for Northern Ireland'
+
+        await forget(request)
+
+        return {
+            'display_region': display_region,
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/start_handlers.py]: Removed

File: app/start_handlers.py [hunk 9]
```diff
@@ -184,41 +184,85 @@ async def post(self, request):
...
+        display_region = 'ni'
+        self.log_entry(request, display_region + '/start/code-for-england')
+
+        locale = 'en'
+        page_title = 'This access code is not part of the census for Northern Ireland'
+
+        await forget(request)
+
+        return {
+            'display_region': display_region,
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/start_handlers.py:243]: Removed

File: app/common_handlers.py [hunk 10]
```diff
@@ -91,6 +91,89 @@ async def get(self, request):
...
+
+        self.log_entry(request, display_region + '/' + user_journey + '/address-in-wales')
+
+        return {
+            'page_title': page_title,
+            'display_region': display_region,
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/common_handlers.py]: Cleaned this up to leave only the needed values

File: app/common_handlers.py [hunk 11]
```diff
@@ -91,6 +91,89 @@ async def get(self, request):
...
+@common_routes.view(r'/ni/' + View.valid_user_journeys + '/address-in-england/')
+class CommonAddressInEngland(CommonCommon):
+    """
+    Route to render an 'Address in England' page during address lookups if display_region is 'ni'
+    and selected addresses region is E
+    """
+    @aiohttp_jinja2.template('common-address-in-england.html')
+    async def get(self, request):
+        self.setup_request(request)
+        display_region = 'ni'
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/common_handlers.py:133]: Cleaned up

File: app/common_handlers.py [hunk 12]
```diff
@@ -91,6 +91,89 @@ async def get(self, request):
...
+            page_title = 'This address is not part of the census for England and Wales'
+            locale = 'en'
+
+        self.log_entry(request, display_region + '/' + user_journey + '/address-in-northern-ireland')
+
+        return {
+            'page_title': page_title,
+            'display_region': display_region,
+            'locale': locale,
+            'user_journey': user_journey,
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/common_handlers.py]: Cleaned up

File: app/start_handlers.py [hunk 13]
```diff
@@ -184,41 +184,85 @@ async def post(self, request):
...
             else:
                 raise HTTPFound(request.app.router['StartConfirmAddress:get'].url_for(display_region=display_region))
         else:
-            if display_region == 'en':
-                raise HTTPFound(request.app.router['StartConfirmAddress:get'].url_for(display_region=display_region))
+            if display_region == 'ni':
+                raise HTTPFound(request.app.router['StartCodeForEngland:get'].url_for())
             else:
-                raise HTTPFound(request.app.router['StartRegionChange:get'].url_for(display_region='en'))
+                raise HTTPFound(request.app.router['StartConfirmAddress:get'].url_for(display_region='en'))
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/start_handlers.py]: Updated

**Review by MartinHumphrey — APPROVED** [review]

File: tests/test_data/address_index/uprn_wales.json [hunk 14]
```diff
@@ -0,0 +1,28 @@
+{
+    "apiVersion": "1.0.0-SNAPSHOT",
+    "dataVersion": "77",
+    "response": {
+        "address": {
+            "uprn": "100100011957",
```
Comment (cookel2) [inline tests/test_data/address_index/uprn_wales.json:6]: looks good

**Review by cookel2 — APPROVED** [review]

**Pull request merged**